### PR TITLE
[improvement](spill) spill trigger improvement

### PR DESCRIPTION
**Pull request by jacktengg**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/spill/spill_writer.cpp [hunk 1]
```diff
@@ -35,13 +35,19 @@ Status SpillWriter::open() {
     return Status::OK();
 }
 
+SpillWriter::~SpillWriter() {
+    if (!closed_) {
+        LOG(WARNING) << "spill writer not closed correctly:\n" << boost::stacktrace::stacktrace();
```
Comment (yiguolei) [inline be/src/vec/spill/spill_writer.cpp:40]: not use this way, boost has a static global lock.
Using doris::Status wrapped error stack, you could use it.

File: be/src/vec/exprs/vexpr.cpp [hunk 2]
```diff
@@ -97,11 +97,11 @@ TExprNode create_texpr_node_from(const void* data, const PrimitiveType& type, in
         break;
     }
     case TYPE_DATEV2: {
-        THROW_IF_ERROR(create_texpr_literal_node<TYPE_DATEV2>(data, &node));
+        THROW_IF_ERROR(create_texpr_literal_node<TYPE_DATEV2>(data, &node, precision, scale));
```
Comment (yiguolei) [inline be/src/vec/exprs/vexpr.cpp:100]: These code is not related with this PR.

**Comment by jacktengg** [comment]
run buildall

**Comment by jacktengg** [comment]
run buildall

**Comment by jacktengg** [comment]
run buildall

File: be/src/pipeline/exec/partitioned_aggregation_source_operator.cpp [hunk 3]
```diff
@@ -128,6 +129,7 @@ Status PartitionedAggSourceOperatorX::close(RuntimeState* state) {
 Status PartitionedAggSourceOperatorX::get_block(RuntimeState* state, vectorized::Block* block,
                                                 bool* eos) {
     auto& local_state = get_local_state(state);
+    local_state.inc_running_big_mem_op_num(state);
```
Comment (yiguolei) [inline be/src/pipeline/exec/partitioned_aggregation_source_operator.cpp:132]: why do we need check local state's count.
Only operator is ok?

File: be/src/pipeline/pipeline_x/pipeline_x_task.cpp [hunk 4]
```diff
@@ -326,6 +304,46 @@ Status PipelineXTask::execute(bool* eos) {
...
+bool PipelineXTask::should_revoke_memory(RuntimeState* state, int64_t revocable_mem_bytes) {
+    auto* query_ctx = state->get_query_ctx();
+    bool need_revoke = query_ctx->query_mem_tracker->need_revoke();
+    if (!need_revoke) {
+        return false;
+    }
+
+    const auto min_revocable_mem_bytes = state->min_revocable_mem();
+
+    auto tg = query_ctx->workload_group();
```
Comment (yiguolei) [inline be/src/pipeline/pipeline_x/pipeline_x_task.cpp:316]: If tg == nullptr then return false; to avoid core when there is no workload group bind to this query. And print log in this case.

Comment (yiguolei) [reply be/src/pipeline/pipeline_x/pipeline_x_task.cpp:316]: do not use tg. tg is old name "task group", use wg

File: be/src/runtime/workload_group/workload_group.h [hunk 5]
```diff
@@ -82,6 +82,31 @@ class WorkloadGroup : public std::enable_shared_from_this<WorkloadGroup> {
...
+
+    void set_weighted_memory_used(int64_t wg_total_mem_used, double ratio);
+
+    std::unordered_set<std::shared_ptr<MemTrackerLimiter>> get_all_mem_trackers();
+
+    int64_t get_weighted_memory_used() {
+        return _weighted_mem_used.load(std::memory_order_relaxed);
+    }
+
+    void check_mem_used(bool& is_low_wartermark, bool& is_high_wartermark) {
```
Comment (yiguolei) [inline be/src/runtime/workload_group/workload_group.h:100]: add const to the function signature.
And do not use reference as return value. Use pointer instead.

File: be/src/pipeline/pipeline_x/pipeline_x_task.cpp [hunk 6]
```diff
@@ -326,6 +304,46 @@ Status PipelineXTask::execute(bool* eos) {
...
+    bool is_tg_mem_low_water_mark = false;
+    bool is_tg_mem_high_water_mark = false;
```
Comment (yiguolei) [inline be/src/pipeline/pipeline_x/pipeline_x_task.cpp:318]: use wg

File: be/src/pipeline/pipeline_x/pipeline_x_task.cpp [hunk 7]
```diff
@@ -326,6 +304,46 @@ Status PipelineXTask::execute(bool* eos) {
...
+    if (tg) {
+        tg->check_mem_used(is_tg_mem_low_water_mark, is_tg_mem_high_water_mark);
+    }
+    if (is_tg_mem_high_water_mark) {
+        if (revocable_mem_bytes > 1024 * 1024) {
```
Comment (yiguolei) [inline be/src/pipeline/pipeline_x/pipeline_x_task.cpp:323]: change this to an session variable.
not use magic numbers.

File: be/src/runtime/memory/mem_tracker_limiter.h [hunk 8]
```diff
@@ -244,6 +245,20 @@ class MemTrackerLimiter final : public MemTracker {
     // Log the memory usage when memory limit is exceeded.
     std::string tracker_limit_exceeded_str();
 
+    void set_weighted_mem(int64_t weighted_limit, double ratio) {
```
Comment (yiguolei) [inline be/src/runtime/memory/mem_tracker_limiter.h:248]: 这个不要在这里搞。
这个信息应该是要记录到query context 中的。

File: be/src/runtime/memory/mem_tracker_limiter.h [hunk 9]
```diff
@@ -244,6 +245,20 @@ class MemTrackerLimiter final : public MemTracker {
...
+        _weighted_consumption = _consumption->current_value() * ratio;
+        _weighted_limit = weighted_limit;
+    }
+    void get_weighted_mem_info(int64_t& weighted_limit, int64_t& weighted_consumption) {
+        std::lock_guard<std::mutex> l(_weighted_mem_lock);
+        weighted_limit = _weighted_limit;
+        weighted_consumption = _weighted_consumption;
+    }
+
+    void set_need_revoke(bool b) { _need_revoke.store(b, std::memory_order_relaxed); }
```
Comment (yiguolei) [inline be/src/runtime/memory/mem_tracker_limiter.h:259]: 这些函数应该都是要再query ctx 中

File: be/src/runtime/workload_group/workload_group.h [hunk 10]
```diff
@@ -141,7 +166,8 @@ class WorkloadGroup : public std::enable_shared_from_this<WorkloadGroup> {
     const uint64_t _id;
     std::string _name;
     int64_t _version;
-    int64_t _memory_limit; // bytes
+    int64_t _memory_limit;                      // bytes
+    std::atomic_int64_t _weighted_mem_used = 0; // bytes
     bool _enable_memory_overcommit;
     std::atomic<uint64_t> _cpu_share;
```
Comment (yiguolei) [inline be/src/runtime/workload_group/workload_group.h:172]: remove query_id_set and use std::map<queryid, weak_ptr<query context>> so that all memory limit info could be set to query context, not to memtracker.

**Comment by jacktengg** [comment]
run buildall

File: be/src/runtime/workload_group/workload_group_manager.cpp [hunk 11]
```diff
@@ -135,6 +137,135 @@ void WorkloadGroupMgr::delete_workload_group_by_ids(std::set<uint64_t> used_wg_i
...
+    int64_t all_queries_mem_used = 0;
+
+    // calculate total memory used of each workload group and total memory used of all queries
+    std::unordered_map<uint64_t, WorkloadGroupMemInfo> wgs_mem_info;
+    for (auto& [wg_id, wg_queries] : all_wg_queries) {
+        int64_t wg_total_mem_used = 0;
+        for (const auto& [query_id, query_ctx_ptr] : wg_queries) {
+            if (auto query_ctx = query_ctx_ptr.lock()) {
+                wg_total_mem_used +=
+                        query_ctx->is_cancelled() ? 0 : query_ctx->query_mem_tracker->consumption();
```
Comment (yiguolei) [inline be/src/runtime/workload_group/workload_group_manager.cpp:165]: 这样可能不行，如果把canceled 设置为0，那么可能会oom，因为此时canceled的query 的内存还没释放。

File: be/src/runtime/workload_group/workload_group_manager.cpp [hunk 12]
```diff
@@ -135,6 +137,135 @@ void WorkloadGroupMgr::delete_workload_group_by_ids(std::set<uint64_t> used_wg_i
...
+                        query_ctx->query_mem_tracker->label(),
+                        query_ctx->query_mem_tracker->parent_label(),
+                        PrettyPrinter::print(query_consumption, TUnit::BYTES),
+                        PrettyPrinter::print(query_weighted_consumption, TUnit::BYTES),
+                        PrettyPrinter::print(query_ctx->query_mem_tracker->peak_consumption(),
+                                             TUnit::BYTES));
+            }
+        }
+        if (wg_mem_info.is_high_wartermark || wg_mem_info.is_low_wartermark) {
+            LOG_EVERY_N(INFO, 3) << debug_msg;
```
Comment (yiguolei) [inline be/src/runtime/workload_group/workload_group_manager.cpp:264]: 改成10 吧

**Comment by jacktengg** [comment]
run buildall

**Review by yiguolei — DISMISSED** [review]

**Comment by jacktengg** [comment]
run buildall

**Review by mrhhsg — APPROVED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]
